samcli/hook_packages/terraform/hooks/prepare/resource_linking.py

```diff
@@ -94,6 +120,68 @@ def _handle_linking(self, source_tf_resource: TFResource, cfn_source_resources:
             source_tf_resource.full_address,
             cfn_source_resources,
         )
+
+        # the config TF resource can map to different CFN resources like in case of using Count to define multiple
+        # resources. This means after apply, each resource can have a different mapped child resources.
+        # see the below example
+        # resource "aws_lambda_function" "function1" {
+        #   count = 2
+        #   layers = ${count.index} == 0 ? [aws_lambda_layer_version.layer1.arn]: [aws_lambda_layer_version.layer2.arn]
+        # }
+        # resource "aws_lambda_layer_version" "layer1" { ... }
+        # resource "aws_lambda_layer_version" "layer2" { ... }
+
+        # It also means that it can happen that some of these resources depends on already applied children resources,
+        # and other resources can depend on some unknown resource like if the customer update the HCL configuration
+        # after applying it, and change one of the source resources to refer to a new child resource.
+
+        # we need to filter out the applied resources and handle them using the actual linking fields mapping approach,
+        # and the non-applied resources, we should use the linking algorithm based on the Config definition of the
+        # resource
+
+        # Filter out the applied and non-applied resources
+        applied_cfn_resources = []
+        non_applied_cfn_resources = []
+        for cfn_resource in cfn_source_resources:
+            linking_field_value = cfn_resource.get("Properties", {}).get(self._resource_pair.cfn_link_field_name)
+
+            # if customer uses any non-applied resources to define the source resource property, Terraform will set
+            # the property value to unknown in the Terraform plan.
+            if linking_field_value is None:
+                non_applied_cfn_resources.append(cfn_resource)
+            else:
+                applied_cfn_resources.append(cfn_resource)
+
+        if applied_cfn_resources:
```

Comment:
do we need these ifs?

Reply:
we need to determine which approach will be used based on the filter output .. so either we can do it here, or in each approach function.

Reply:
if the lists are empty, we won't go through the loop. so we can remove those ifs.

Reply:
this can work for the applied approach, and I update that, but for the non applied, we do some other logic before we loop on the CFN resource to do the mapping, and I do not want to do some non required processing. I can move the other if condition to be done in the beginning of the `_link_using_terraform_config`